Fixes #29582 - Improve performance of externalNodes

lzap opened the pull request:
This patch speeds up externalNodes action by 13 times: from 14 ops/s to 190 ops/s. You can test this on your setup, benchmark is included - it tests ENC with a host with 1000 facts and 1000 host parameters with a simple ERB block. To run it (I used Foreman develop with PostgreSQL from Fedora 31):

```
RAILS_ENV=production be ruby -Itest test/benchmark/host_info_benchmark.rb
```

Here is how I was researching the code and implementing changes and the gain I got from the change.

* Initial performance: 14 ops/second
* Removed deep_stringify_keys call: 14 ops/second
* Hash is passed as argument (less top level copying) 22 ops/second
* Skip ERB rendering when string does not contain `<%` 28 ops/second
* Memorized renderer scope 28 ops/second
* Moved interpolate_erb_in_parameters settings check out of the loop: 184 ops/second
* Introducing inplace parse!: 189 ops/second

So the biggest speedup is by simply moving this line out of the private method which is called many times to the entry method:

```
  return value unless (Setting[:interpolate_erb_in_parameters])
```

I do believe that our settings stack is ultra slow and if we refactor it in a way that a setting lookup is very fast it would boost other parts of Foreman too.

tbrisker commented on app/controllers/hosts_controller.rb:
```diff
@@ -193,10 +193,7 @@ def externalNodes
-        # Remove the HashesWithIndifferentAccess using 'deep_stringify_keys',
-        # then we turn it into YAML
-        host_info_yaml = @host.info.deep_stringify_keys.to_yaml(:line_width => -1)
+        host_info_yaml = @host.info.to_yaml(:line_width => -1)
```
You mentioned this didn't have a significant impact, so i think we should skip it - i believe it is needed to make sure that hash keys are properly rendered in the enc

ohadlevy replied:
just a note of caution, if people ENC output will be different, that should be a breaking API change, please note changes in ENC output effects system configurations! afaiu deep merge etc are all there for a reason?

lzap replied:
Yes, I was planning to drop it but I wanted to bring it to reviewers attention. Thing is - why we do have this? There is a note this is for 1.9.x version which is unsupported now. I also did compare my output with and without this method and I see no difference. So why is that?

ekohl replied:
If it's a minor performance factor, perhaps it makes sense to send it as a separate PR so you don't have multiple discussions going on at once? https://theforeman.org/2019/03/merge-time-review-complexity.html

lzap replied:
Someone explain to me why this method should be there? In YAML there is no difference between symbol and string, is there?

```
[3] pry(main)> {test: 1}.to_yaml
=> "---\n:test: 1\n"
[4] pry(main)> {'test': 1}.to_yaml
=> "---\n:test: 1\n"
```

> just a note of caution

I understand, test coverage of this action is rather poor.

adamruzicka replied:
There is a difference, but you use symbols as keys in both of the examples before it ever gets to yaml

```ruby
{test: 1}.keys.first.class
=> Symbol

{'test': 1}.keys.first.class
=> Symbol

{:test => 1}.keys.first.class
=> Symbol

{"test" => 1}.keys.first.class
=> String
```

lzap replied:
I have just found we actually have tests for this controller, I was searching in wrong directory. Now I understand it, I will put this call back:

```
Rails logs for test_0016_externalNodes should render YAML hashes correctly FAILURE:
HostsControllerTest#test_0016_externalNodes should render YAML hashes correctly = 0.19 s = F


Failure:
HostsControllerTest#test_0016_externalNodes should render YAML hashes correctly [test/controllers/hosts_controller_test.rb:257]:
--- expected
+++ actual
@@ -81,9 +81,11 @@
 classes:
   dhcp:
     bootfiles:
-    - name: foo
+    - !ruby/hash:ActiveSupport::HashWithIndifferentAccess
+      name: foo
       mount_point: \"/bar\"
-    - name: john
+    - !ruby/hash:ActiveSupport::HashWithIndifferentAccess
+      name: john
       mount_point: \"/doe\"
 environment: production
```

lzap replied:
Yeah thanks Adam, I forgot that YAML supports these nasty things. Okay, makes sense, I am putting this back.

tbrisker commented on app/services/parameter_safe_render.rb:
```diff
@@ -6,18 +6,36 @@ def initialize(host)
     @host = host
   end
```
Would it make sense to include 
`
@interpolate = Setting[:interpolate_erb_in_parameters]
`
in the initializer so it is only looked up once (assuming the same renderer is reused) and not on every render?

lzap replied:
Good idea, one renderer is actually shared but only N times (constant). Won't hurt.

tbrisker commented on app/services/parameter_safe_render.rb:
```diff
@@ -6,18 +6,36 @@ def initialize(host)
     @host = host
   end
 
+  # Faster method that DOES not modify the input
```
I think you mean 
```suggestion
  # Faster method that DOES modify the input
```

lzap replied:
Yes, yes.

tbrisker reviewed (changes requested):
Looks very promising! see inline comments

lzap commented:
Resolved rubocop, removed the mentioned "optimalization". Luckily not affected, we are still getting over 10x performance boost. I'd appreciate testing with some real data, perhaps @ekohl could hotfix his instance? This can affect Puppet users a lot (in a bad way too).

ekohl commented on app/controllers/hosts_controller.rb:
```diff
@@ -193,9 +193,6 @@ def externalNodes
     begin
       respond_to do |format|
-        # don't break lines in yaml to support Ruby < 1.9.3
-        # Remove the HashesWithIndifferentAccess using 'deep_stringify_keys',
-        # then we turn it into YAML
```
This part is still relevant

lzap replied:
It is but we don't put comments on every line explaining what it does, do we? ;-)

I felt like it's not needed.

ekohl replied:
Well, you yourself in this PR missed why it was useful so clearly it's not obvious.

ekohl commented on app/services/parameter_safe_render.rb:
```diff
@@ -6,18 +6,40 @@ def initialize(host)
-  def render_object(object)
-    return object unless (Setting[:interpolate_erb_in_parameters])
+  def interpolate_erb?
+    @interpolate_erb ||= Setting[:interpolate_erb_in_parameters]
```
The caching doesn't work if the value is false.

lzap replied:
Good call, but it does work when it's true and it is by default. Shall I keep it this way?

lzap replied:
I lean towards putting it back (no memorizing), we only use it at a single place approx. four times.

ekohl replied:
@tbrisker suggested to set the value in the initializer which I think is good enough.

lzap replied:
There is actually a way to memorize `false` values! Let me do it.

tbrisker commented on app/models/concerns/host_info_extensions.rb:
```diff
@@ -23,11 +23,11 @@ module HostInfoExtensions
       provider = provider_class.new(self)
-      info = provider.host_info
-      info_hash.deep_merge! info if info
+      provider.host_info!(info_hash)
```
What about info providers not in core?

lzap replied:
I did only find katello. My plan is to file PR.

lzap replied:
Oh Ansible has that too.

lzap replied:
I can deprecate the old method too. But that requires a bit logic around detecting what provider supports.

timogoebel replied:
we also use the info provider API in (internal) plugins.

lzap replied:
Implemented a detection and deprecated warning.

lzap commented:
TODO: Once approved and tested file PRs for:

* [ ] Katello
* [ ] Ansible

lzap commented:
Rebased.

ares commented on app/services/parameter_safe_render.rb:
```diff
@@ -26,14 +49,19 @@ def render_object(object)
   end
 
+  def renderer_scope
+    @renderer_scope ||= Foreman::Renderer.get_scope(klass: Foreman::Renderer::Scope::Partition, host: host)
```
I would neee to check, but doesn't the scope holds e.g. template name? Does this work for rendering snippets correctly? Don't they have different variables also?

timogoebel replied:
I'm wondering if `Partition` is the right scope?

lzap replied:
Good questions, wrong place :-)

If I am touching this line, I will only memorize this for now.

ares reviewed (commented):
From what I read, the most important optimization is just one of the changes here. I'd prefer to split this up to small commits. This touches many areas so it's more time consuming to test. Touching rendering and host info providers in single change is not good. It will be harder, if it turns out we need to revert some part. Also small changes are easier to cherrypick.

What if we start with just changing Settings?

ares commented:
One more thought, is this setting even useful? If we trust the safemode, why would someone disable this?

And @lzap please don't get me wrong, this is fantastic finding and effort. There are many areas that would benefit greatly from such walk through!

ekohl commented:
> One more thought, is this setting even useful? If we trust the safemode, why would someone disable this?

Performance? Reliability? Maybe a module accepts an ERB template so you don't want rendering on the Foreman side (due to scoping).

> I'd prefer to split this up to small commits.

:+1: to that. Given some cause a deprecation while others don't, some should be much cheaper to merge. It can also be a better example of what modifications are needed in a plugin when the change is small.

ares commented:
Ok, I don't think there woupd be big performance impact, based on what lzap found. Not sure about reliability. By specifying erb, I explicitly say this can change and will be always calculated on demand.

However the parameter expecting erb sounds as a good justification. It can be escaped using `<%%` however as a user, I don't want to escape all my templates.

Still I wonder how many people know and change this setting. I'm tempted to ask on discourse :-) but given above, that's not important for this PR.

tbrisker commented:
I wonder if we'll see the same impact in production with the setting change. Setting lookups should be cached: https://github.com/theforeman/foreman/blob/develop/app/models/setting.rb#L94-L99 so it shouldn't have such an impact.

timogoebel commented:
> Setting lookups should be cached

Do we know if the self-developed caching that loads the data from disk/redis/memcache is faster than the per-query caching ActiveRecord already does? Maybe it would be faster to preload all data in a single query (`select * from settings`) then loading the individual settings (n times `select * from settings where name = '...'`).
Isn't this where we could gain a lot of speed all over the app?

lzap commented:
Rebased.

> And @lzap please don't get me wrong, this is fantastic finding and effort. There are many areas that would benefit greatly from such walk through!

I considered the same. As I was testing more and more changes I went too wild to be honest. To summarise, these changes I think are good and safe to merge:

* Moved interpolate_erb_in_parameters settings check out of the loop
* Skip ERB rendering when string does not contain <%
* Memorized renderer scope (on the edge, little perf benefit)

These are too much of a change, dangerous and incorrect in one case:

* Removed deep_stringify_keys call: 14 ops/second
* Hash is passed as argument (less top level copying) 22 ops/second
* Introducing inplace parse!: 189 ops/second

How about such a split? The patch will be very small and even backportable :-)

lzap commented:
> Do we know

No.



> Isn't this where we could gain a lot of speed all over the app?

Yes, for those cases where we use Setting in a loop.

I believe that settings should be loaded in a single query into memory just like you propose and kept in a hash. This hash should be only invalidated on write to settings. Basically a cache but custom made.

I think that connecting to redis or reading a file is fast enough to do once or twice in a request but too slow operation to do this 1000 times in a request. That was exactly the case. I'd be VERY surprised if this was the only case.

Edit: By the was we do the same thing with permissions and we do bunch of permissions check per request. Probably less than thousands tho.

timogoebel commented:
> Yes, for those cases where we use Setting in a loop.

I also meant if we can reduce the number of individual queries per request.

lzap commented:
> I also meant if we can reduce the number of individual queries per request.

Yes, makes a lot of sense. Should not be too difficult to do actually. :-)

lzap committed:
Fixes #29582 - Improve performance of externalNodes

lzap commented:
Okay gents, 6 lines change before:

```
Warming up --------------------------------------
            Thousand     1.000  i/100ms
Calculating -------------------------------------
            Thousand     13.511  (±14.8%) i/s -    127.000  in  10.081058s
```

and after:

```
Warming up --------------------------------------
            Thousand    14.000  i/100ms
Calculating -------------------------------------
            Thousand    180.137  (±17.2%) i/s -      1.708k in  10.023687s
```

The only thing which can ruin my day is finding out I was actually testing some edge case and there is no gain on production setup.

tbrisker commented on app/services/parameter_safe_render.rb:
```diff
@@ -26,12 +30,14 @@ def render_object(object)
 
   def render_string(string)
+    # exit early if there is nothing to parse (minimum possible ERB is four characters '<%%>')
+    return string if (string.nil? || string.size <= 4 || !string.index('<%'))
```
we have `String#contains_erb?` in `lib/core_extentions.rb` - might be good to use that (and improve it if possible).

lzap replied:
Will do, our implementation is not the fastest :-)

tbrisker replied:
also, `string != nil` since this function is only called when the object is a `String`

lzap replied:
Well not anymore when I extract this into `contains_erb?` I guess.

tbrisker replied:
it is because `contains_erb?` is monkey patched to String class. So not crucial, but the nil check is redundant both in this function and in that one.

lzap replied:
Oh now I got what you mean, sure.

lzap commented:
Rebased, those failing tests seem to be JS only hopefully the new reabase solves them.

lzap commented:
For the record, I was testing with Redis cache store. I created a small bench for major Rails cache implementations and the preformance is  roughly the same:

```
Warming up --------------------------------------
    memory_store hit     1.000  i/100ms
   memory_store miss     1.000  i/100ms
Calculating -------------------------------------
    memory_store hit      2.852k (±14.7%) i/s -     11.440k in   4.991405s
   memory_store miss      2.860k (±13.7%) i/s -     11.679k in   4.989070s
Warming up --------------------------------------
      file_store hit     1.000  i/100ms
     file_store miss     1.000  i/100ms
Calculating -------------------------------------
      file_store hit      2.805k (±18.3%) i/s -     11.037k in   4.991170s
     file_store miss      2.867k (±14.0%) i/s -     11.554k in   4.990827s
Warming up --------------------------------------
redis_cache_store hit
                         1.000  i/100ms
redis_cache_store miss
                         1.000  i/100ms
Calculating -------------------------------------
redis_cache_store hit
                          2.897k (±14.2%) i/s -     11.618k in   4.991224s
redis_cache_store miss
                          2.866k (±13.9%) i/s -     11.707k in   5.056739s
```

Something bad is in our Settings implementation.

```ruby
require 'benchmark/ips'

task benchmark: :environment do
  stores = {
    memory_store: [{ size: 64.megabytes }],
    file_store: [Rails.root + "/tmp/cache"],
    redis_cache_store: [{url: ["redis://localhost:8479/0"]}],
  }

  actions = {
    :hit => lambda {
      Rails.cache.fetch("test")
    },
    :miss => lambda {
      Rails.cache.read("test"+rand().to_s[0..10])
    }
  }

  stores.each do |store,args|
    ActionController::Base.cache_store = store, *args

    Benchmark.ips do |x|
      x.config(:time => 5, :warmup => 0)

      actions.each do |label, proc|
        Rails.cache.delete("test")
        Rails.cache.fetch("test") { [Time.now, 1.year.ago] }
        x.report("#{store} #{label}") do
          proc.call
        end
      end
    end
  end
end
```

timogoebel commented:
> Something bad is in our Settings implementation.

What I meant was, that ActiveRecord calls are automatically cached in memory and don't hit the db more than one time anyway. Do we need a custom cache implementation using `Rails.cache`?

lzap commented:
Guys, I have some bad news to share. I have found that I had misconfiguration of my Redis, therefore all writes were actually returning `false`, therefore there was no caching effectively.

I am going to rerun the benchmark now with caching, but I am afraid there will be zero performance impact.

This points out on an important thing tho, Rails caching does not error out and you can silently have a deployment without any caching. We should implement a simple check during boot via an initializer and error out if caching does not work as this can dramatically improve performance as we can see.

timogoebel commented:
> This points out on an important thing tho, Rails caching does not error out and you can silently have a deployment without any caching. We should implement a simple check during boot via an initializer and error out if caching does not work as this can dramatically improve performance as we can see.

That makes a lot of sense. Don't we have some mechanism already to check the state of "external" services (like e.g. Candlepin)?

ekohl commented:
> That makes a lot of sense. Don't we have some mechanism already to check the state of "external" services (like e.g. Candlepin)?

There is the ping API command. That could do a set(time()) + get to verify it's working.

lzap commented:
Honestly I believe this should be done during initialization, unless you do the ping commend (which you don't have to) you are running in the "slower mode".

timogoebel commented:
> unless you do the ping commend (which you don't have to) you are running in the "slower mode".

I honestly believe we should do both.

lzap commented:
> I honestly believe we should do both.

Can do, okay.

Ok, this is weird. When I switch cache back to file_cache (which is about the same performance as Redis), I still see huge improvement! Even bigger now, uh. Someone help me to understand this :-)

Develop:

```
Warming up --------------------------------------
            Thousand    21.000  i/100ms
Calculating -------------------------------------
            Thousand    226.408  (±15.0%) i/s -      2.205k in  10.039612s
```

This PR:

```
Warming up --------------------------------------
            Thousand   125.000  i/100ms
Calculating -------------------------------------
            Thousand      1.541k (±12.7%) i/s -     15.125k in  10.028064s
```

ekohl commented:
> I honestly believe we should do both.

Indeed. Redis can go offline and ping should report on that.

ekohl commented:
Perhaps file caches have a larger overhead than the database for a single query? Note that in production our file caches are in `/run` (yes, not the best place) so a tmpfs.

lzap commented:
> Perhaps file caches have a larger overhead than the database for a single query?

Not at all, file store is probably one of the fastest cache implementations. I think fetching a key from a file store is faster than even finishing connecting to PostgreSQL over TCP.

> Note that in production our file caches are in `/run` (yes, not the best place) so a tmpfs.

I've seen wonderful talk about tmpfs vs normal file system and they revealed it's the very same performance (plus you have persistance). It's because how Linux caches file operations.

lzap commented:
Do I read this right that develop is at `2.205k` and PR at `15.125k` so this is still like 8x improvement?

lzap commented:
Added: https://github.com/theforeman/foreman/pull/7607

This PR is still valid, if anyone can run the benchmark and confirm the findings that would be great.

lzap commented:
Anyone see still improvement as me, even with proper caching? Reports from production deployment are not good, minimum to none improvements :-(

tbrisker reviewed (approved):
This is still a valid change even if the outcome isn't as large as initially identified. The benchmark can help identify further areas for improvement.

lzap commented:
For the record, the patch helped a lot for a community user, there is a nice graph available with the difference:

https://community.theforeman.org/t/foreman-2-0-0-memory-leak/18953/23

![04969e8cad50211ed64763d0e2b36825ebee77f0](https://user-images.githubusercontent.com/49752/83498155-cfc73500-a4bb-11ea-9aa7-fc316516085b.png)

ekohl commented:
@tbrisker should we pull this into 2.0? It looks like a safe fix

tbrisker commented:
2.0 - a588bd34a
